[Draft] [D3D] Support variable refresh rate displays (vsync off) in borderless windowed

duckdoom4 opened the pull request:
I made an attempt to improve support for vsync off in borderless windowed mode. (See https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/variable-refresh-rate-displays)

The [documentation recommends](https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/for-best-performance--use-dxgi-flip-model) we no longer use `DXGI_SWAP_EFFECT_DISCARD` and switch to `DXGI_SWAP_EFFECT_FLIP_*` instead. 

Still TODO:
- [ ] MSAA is not supported in this model. The dx11 implementation will need to be updated to support this. There is a detailed explaination on how to resolve that [here](https://learn.microsoft.com/en-us/windows/win32/api/dxgi/ne-dxgi-dxgi_swap_effect#remarks). (Would love your help with this one, I assume we can use an almost exact copy of the dx12 implementation?)
- [x] Disable tearing if fullscreen exclusive mode is active
- [x] Verify the code will work under different configurations
- [ ] Update the dx11 code so it matches the dx12 version? (Aka. use device to create swap chain, and move tearing support check to device as well)

Let me know what you think, would love your feedback :)

LukasBanana reviewed (changes requested):
I'm happy to see this project getting modernized with features like variable refresh rate.

With respect to C++17/C++20 features: I want to keep compatibility strictly with C++11. My rationale behind this is to stay compatible with as many platforms and legacy systems as possible. Supporting legacy systems is basically only intended for the GL backend, but I like to have a cohesive code base that uses the same language standard throughout. Having said that, I don't mind adding some of those features like `[[maybe_unused]]` to the [CompilerExtensions.h](https://github.com/LukasBanana/LLGL/blob/master/sources/Core/CompilerExtensions.h) file like so:
```cpp
#if __cplusplus >= 202002L // C++20
#   define LLGL_MAYBE_UNUSED [[maybe_unused]]
#else
#   define LLGL_MAYBE_UNUSED
#endif
```
With the macro that's already defined there I made an exception with my naming guideline so you could write `if unlikely(X) {}`.

duckdoom4 commented on sources/Renderer/Direct3D12/D3D12SwapChain.cpp:
```diff
@@ -103,7 +104,10 @@ void D3D12SwapChain::SetDebugName(const char* name)
 {
     /* Present swap-chain with vsync interval */
-    HRESULT hr = swapChainDXGI_->Present(syncInterval_, 0);
+    bool tearingEnabled = tearingSupported_ && syncInterval_ == 0;
```
Ah, i thought dx12 didn't support the same full screen exclusive mode as dx11. But if you did implement that, then yes, it would need the same check

duckdoom4 commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -71,7 +75,9 @@ void D3D11SwapChain::SetDebugName(const char* name)
 void D3D11SwapChain::Present()
 {
-    swapChain_->Present(swapChainInterval_, 0);
+    bool tearingEnabled = tearingSupported_ && swapChainInterval_ == 0; // TODO: disable if fullscreen exclusive mode is enabled
```
Yeah, I think so. The documentation didn't mention a way to do this so I'd need to do some googling to see if there is some (easy) way to check

duckdoom4 replied:
I checked out this example they have on github, which also contains the tearing support. https://github.com/microsoft/DirectX-Graphics-Samples/blob/master/Samples/Desktop/D3D12Fullscreen/src/D3D12Fullscreen.cpp

It seems all they do there is test `GetFullScreenState` in the swapchain resize function and store that in the `m_windowedMode` bool. I can do the same here. 

They also have a separate handler for alt+enter replacing it with spacebar, but if you look at the window message that handles the space press the comment suggests that alt+enter does the same thing. So I guess that's not a requirement. (Also didn't see any mention of this in any other documentation so far)

duckdoom4 commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -260,19 +266,45 @@ static UINT GetPrimaryDisplayRefreshRate()
 
     /* Find suitable multi-samples for color format */
-    swapChainSampleDesc_ = D3D11RenderSystem::FindSuitableSampleDesc(device_.Get(), colorFormat_, samples);
+    swapChainSampleDesc_ = D3D11RenderSystem::FindSuitableSampleDesc(device_.Get(), colorFormat_, 1); // TODO: resolve multi-sampling
```
Sounds good. I think I'll have some time on Monday so I can finish it then

duckdoom4 commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -381,6 +418,23 @@ void D3D11SwapChain::RestoreDebugNames(const std::string (&debugNames)[4])
     }
 }
 
+void D3D11SwapChain::CheckTearingSupport([[maybe_unused]] IDXGIFactory* factory)
```
Ah right, I forgot to put it back in c++11 mode after I made the MR to fix the c++20 issues I had in my own project. I'll put it back so I don't forget in the future

LukasBanana commented:
Changes look good at first glance. I'll take a closer look and test this on Thursday.

LukasBanana commented on sources/Renderer/Direct3D12/D3D12SwapChain.cpp:
```diff
@@ -441,13 +448,17 @@ HRESULT D3D12SwapChain::CreateResolutionDependentResources(const Extent2D& resol
         swapChain.As(&swapChainDXGI_);
     }
 
+    BOOL fullscreenState;
```
Unless it's a larger buffer that is intended to be uninitialized at allocation time (like `LLGL::DynamicByteArray{ N, LLGL::UninitializeTag{} }` for instance), I'd prefer to have all variables default initialized. Skipping the `= FALSE` here is really just a micro optimization.

duckdoom4 replied:
Ah yeah missed that, my bad

LukasBanana commented on sources/Renderer/Direct3D12/D3D12RenderSystem.cpp:
```diff
@@ -19,19 +19,19 @@
 
 #include <LLGL/Backend/Direct3D12/NativeHandle.h>
 
+#include <dxgi1_5.h>
```
What minimum Windows SDK version requires DXGI 1.5? So far DXGI 1.4 was the highest used DXGI version in the D3D12 backend, which I set the WinSDK version to 10.0.10240.0 in the [docs page](https://github.com/LukasBanana/LLGL/tree/master/docu#windows). Would this require to lift the SDK version or can we wrap this into an `#if LLGL_D3D12_ENABLE_FEATURELEVEL` block like with the D3D11 backend? Having that said, I think I'll have to split this macro up anyway, since feature level, D3D version, DXGI version, and shader-model are all separate concepts that are not tied together 1:1, but right now all of these things are wrapped under the same CMake option and macro.

duckdoom4 replied:
According to [this page](https://github.com/toivjon/dxgi-sandbox), it's Windows 10.

LukasBanana commented on sources/Renderer/Direct3D12/D3D12RenderSystem.cpp:
```diff
@@ -19,19 +19,19 @@
 
-#include "RenderState/D3D12GraphicsPSO.h"
 #include "RenderState/D3D12ComputePSO.h"
+#include "RenderState/D3D12GraphicsPSO.h"
```
I presume this change isn't really necessary.

duckdoom4 replied:
True, that was my auto formatter sorting the includes

LukasBanana commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -260,18 +267,49 @@ static UINT GetPrimaryDisplayRefreshRate()
+    }
+
+    ComPtr<IDXGIFactory2> factory2;
+    HRESULT hr = factory->QueryInterface(IID_PPV_ARGS(&factory2));
```
Just for the factory query, I think it makes sense to move this upward within this `#if` block and if it fails, fallback to the alternative swap-chain creation, just like you did in `D3D11SwapChain::CheckTearingSupport`. If `CreateSwapChainForHwnd` fails with a valid factory, we should indeed crash with a callstack.

duckdoom4 replied:
Makes sense, will do that

LukasBanana commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -260,18 +267,49 @@ static UINT GetPrimaryDisplayRefreshRate()
+    swapBuffers = std::max(2u, std::min<std::uint32_t>(swapBuffers, DXGI_MAX_SWAP_CHAIN_BUFFERS));
+
+    /* Find suitable multi-samples for color format */
+    swapChainSampleDesc_ = D3D11RenderSystem::FindSuitableSampleDesc(device_.Get(), colorFormat_, 1); // TODO: resolve multi-sampling
```
I can take care of adopting the same back-buffer creation as in D3D12.

LukasBanana commented on sources/Renderer/Direct3D11/D3D11RenderSystem.cpp:
```diff
@@ -551,8 +551,25 @@ void D3D11RenderSystem::ClearStateForAllContexts()
-    HRESULT hr = CreateDXGIFactory(IID_PPV_ARGS(factory_.ReleaseAndGetAddressOf()));
+#if LLGL_D3D11_ENABLE_FEATURELEVEL >= 2
+
+    HRESULT hr = CreateDXGIFactory2(0, IID_PPV_ARGS(&factory2_));
```
These should also all fallback to the next lower factory version if they fail. `LLGL_D3D11_ENABLE_FEATURELEVEL` is only a guard to let LLGL compile on systems with older Windows SDKs. It shouldn't require that the runtime version supports anything higher than Direct3D 11.0.

LukasBanana reviewed (changes requested):
I left a few comments. With the SDK requirement for DXGI 1.5, I'm okay with leaving that as-is since the solution with the `LLGL_D3D12_ENABLE_FEATURELEVEL` macro is not optimal right now anyway. If you can adjust the other points like making sure we can fall back to lower versions of the factory if the runtime doesn't support it (regardless of what's enabled via the feature level macros) I'll merge this into the master branch and adjust the D3D11 backend to support multi-sampled backbuffers.

duckdoom4 commented:
Had a bussy week, but got the comments addressed now.

Added the fallback to the next lower factory version on failure, and added the fallback to the next swap chain version on failure by adding a new function to make it more readable

duckdoom4 commented on sources/Renderer/Direct3D11/D3D11RenderSystem.cpp:
```diff
@@ -671,9 +692,10 @@ void D3D11RenderSystem::QueryDXDeviceVersion()
 {
     LLGL_ASSERT_PTR(device_);
 
+    HRESULT hr;
```
Moved this here, because line 681/703 would fail to compile if `LLGL_D3D11_ENABLE_FEATURELEVEL`  is set to `2`

duckdoom4 commented on sources/Renderer/Direct3D11/D3D11RenderSystem.cpp:
```diff
@@ -671,9 +692,10 @@ void D3D11RenderSystem::QueryDXDeviceVersion()
 {
     LLGL_ASSERT_PTR(device_);
 
+    HRESULT hr = S_OK;
```
Moved this here, because line 681/703 would fail to compile if LLGL_D3D11_ENABLE_FEATURELEVEL is set to 2

duckdoom4 replied:
Moved this here, because line 681/703 would fail to compile if LLGL_D3D11_ENABLE_FEATURELEVEL is set to 2

LukasBanana commented on sources/Renderer/Direct3D11/D3D11SwapChain.cpp:
```diff
@@ -286,9 +305,45 @@ void D3D11SwapChain::CreateSwapChain(IDXGIFactory* factory, const Extent2D& reso
     }
-    HRESULT hr = factory->CreateSwapChain(device_.Get(), &swapChainDesc, swapChain_.ReleaseAndGetAddressOf());
+
+    hr = factory->CreateSwapChain(device_.Get(), &swapChainDesc, swapChain_.ReleaseAndGetAddressOf());
```
This one fails with `'hr': undeclared identifier` if `LLGL_D3D11_ENABLE_FEATURELEVEL` is 2 or lower (see [AppVeyor CIS job](https://ci.appveyor.com/project/LukasBanana/llgl/builds/49605941/job/3undo0l8qxus4w7e#L257)). Can likely be fixed the same way you fixed it in _D3D11RenderSystem.cpp_.

LukasBanana reviewed (changes requested):
Just one compile error to fix and we're good to go. I'm also a bit slow on responding right now, because I'm on vacation and will be back on Thursday.

LukasBanana reviewed (approved):
Looks good. Thanks again for your contribution. I will add the MSAA support for D3D11 next weekend.

duckdoom4 commented:
Yeah you're welcome. Happy to help. I'm actually looking at adding some other features as well so might make some more in the near future.

For example adding support for HDR framebuffer output for HDR displays.

Also I need bindless textures for a research project I'm working on. 
Does LLGL support that currently? I already read that dx11 does not support that,since it's a shader model 5.1 thing

LukasBanana commented:
Added support for multi-sampled back-buffers with the FLIP swap-effect with 5778b34.
